### PR TITLE
Enable to write the nextlink for the collection of entity reference links

### PR DESCRIPTION
**Pull request by xuzhg**
…inks

### Issues

*This pull request fixes issue #xxx.*

### Description

* In the Response payload, we should allow to write the next link for the collection of entity reference links.

This PR is supporting this.

Type | version | Request | Response
-- | -- | -- | --
Single Navigation Property | 4.0 | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReport@odata.bind":   "http://host/service/Employees(5)"   } | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReport   ": {        \"@odata.id\":\"Orders(2)\"     }   }
Single Navigation Property | 4.01 | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReport":   {   "@id": "Employees(5)"     }   } | {       "@odata.type":"#Northwind.Manager",     "DirectReport":   {         "@id": "Orders(5)"     }   }
Collection Navigation Property | 4.0 | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReports@odata.bind":   [         "http://host/service/Employees(5)",         "http://host/service/Employees(6)"     ]   } | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReports   ": [        {\"@odata.id\":\"Orders(2)\"},        {\"@odata.id\":\"Orders(3)\"}     ]   }
Collection Navigation Property | 4.01 | {       "@odata.type":"#Northwind.Manager",       "ID": 1,     "DirectReports":   [         {"@odata.id": "Employees(5)"},         {"@odata.id": "Employees(6)"}     ]   } | {       "@odata.type":"#Northwind.Manager",     "DirectReports":   [         {"@id": "Orders(5)"},         {"@id": "Orders(6)"}     ]   }

# 05/21/2019 updates

We should make the request and response writing with the same pattern, and avoid breaking changes in the request of collection nested resource by enabling the following patterns, 

### Old pattern in request
```
WriteStart( nestedResourceInfo ) // collection
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
WriteEnd() // end of nested resource info
```
should remove the old pattern in the next breaking change version.

### New pattern in request (Need more discussion)
```
WriteStart( nestedResourceInfo ) // collection
   WriteStart( new ODataResourceSet() ) // Add a resource set writing logic
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
  WriteEnd() // end of resource set
WriteEnd() // end of nested resource info
```

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 1]
```diff
@@ -1018,32 +1021,51 @@ protected override void WriteEntityReferenceInNavigationLinkContent(ODataNestedR
             // In JSON Light, we can only write entity reference links at the beginning of a navigation link in requests;
             // once we wrote a resource set, entity reference links are not allowed anymore (we require all the entity reference
             // link to come first because of the grouping in the JSON Light wire format).
-            JsonLightNestedResourceInfoScope navigationLinkScope = (JsonLightNestedResourceInfoScope)this.CurrentScope;
-            if (navigationLinkScope.ResourceSetWritten)
+            JsonLightNestedResourceInfoScope nestedResourceScope = this.CurrentScope as JsonLightNestedResourceInfoScope;
+            if (nestedResourceScope == null)
+            {
+                nestedResourceScope = this.ParentNestedResourceInfoScope as JsonLightNestedResourceInfoScope;
```
Comment (KanishManuja-MS) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: >ParentNestedResourceInfoScope [](start = 43, length = 29)

This is not clear to me; if a parentNestedResourceInfo is passed, why are we using the private member? Should the names be different then?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 2]
```diff
@@ -990,7 +990,10 @@ protected override void EndNestedResourceInfoWithContent(ODataNestedResourceInfo
             // resource set afterwards, we have to now close the array of links.
             if (navigationLinkScope.EntityReferenceLinkWritten && !navigationLinkScope.ResourceSetWritten && nestedResourceInfo.IsCollection.Value)
             {
-                this.jsonWriter.EndArrayScope();
+                if (!this.jsonLightOutputContext.WritingResponse)
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: why make this a separate if, and not part of the above condition (perhaps with line breaks inserted for readability)?

Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Why does this depend on whether or not we're writing a response. There are valid requests that take an array of entity references, enclosed in square brackets -- are the square brackets for that case written somewhere else?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 3]
```diff
@@ -1018,32 +1021,51 @@ protected override void WriteEntityReferenceInNavigationLinkContent(ODataNestedR
...
+            if (!nestedResourceScope.EntityReferenceLinkWritten)
             {
-                if (!this.jsonLightOutputContext.WritingResponse &&
-                    (this.Version == null || this.Version < ODataVersion.V401))
-                {
-                    // Write the property annotation for the entity reference link(s)
-                    this.odataAnnotationWriter.WritePropertyAnnotationName(parentNestedResourceInfo.Name, ODataAnnotationNames.ODataBind);
-                }
-                else
+                // In request
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: >In request [](start = 19, length = 10)

I understand writing a different annotation in a request for 4.0 versus 4.01 and greater, but not sure why the logic for writing the open/close square bracket varies from request to response.

**Comment by mikepizzo** [comment]
Might help to meet quickly to discuss. I'm unclear why the logic for writing square brackets in collections that contain entity references differs from request to response. If this is because we have some legacy code pattern that writes a list of entity references, then that pattern should take care of writing the []; in the normal pattern of writing the nested info and then writing either a resource or resourceset, writing the nested resource set should always take care of the open/close brackets.

**Comment by mikepizzo** [comment]
The only solution I can come up with for writing both the old (4.0) and new (4.01) syntax and still support the old API is to defer writing the property name until the first element, even for the case of a nested collection following the new pattern (i.e., the property name would not be written on WriteStart(ResourceSet) but when writing the first resource, or on WriteEnd() for the resourceSet if no resource were written.)

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: @KanishManuja-MS What do you mean? 

ParentNestedResourceInfoScope is used if there's a ResourceSet.

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 3 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: in Request:
```
WriteStart( nestedResourceInfo ) // collection
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
WriteEnd() // end of nested resource info
```
the open square bracket is written in first call to : "WriteEntityReferenceLink(..)

Meanwhile, in response, 

```
WriteStart( nestedResourceInfo ) // collection
   WriteStart( new ODataResourceSet() ) // Add a resource set writing logic
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
      WriteEntityReferenceLink( EntityReferenceLink)
  WriteEnd() // end of resource set
WriteEnd() // end of nested resource info
```

the open square bracket is written in the "WriteStart(new ODataResourceSet()), so, when we call the first "WriteEntityReferenceLink(...)", it needn't to write the open square bracket. 

Same for the close square bracket.



---
In reply to: [285803927](https://github.com/OData/odata.net/pull/1464#discussion_r285803927) [](ancestors = 285803927)

**Comment by xuzhg** [comment]
defer the property name and open square bracket until the first element maybe a breaking change for current behavior and it's a lit complicated, because it's not only related to the property name and open square bracket, but also other instance annotations. 

---
In reply to: [497369236](https://github.com/OData/odata.net/pull/1464#issuecomment-497369236) [](ancestors = 497369236)

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 2 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Yes. Writing the close square bracket in response is in `WriteEnd()` for resource set. We don't need to write `[` again here. 

Besides, `WriteEnd()` for resource set will process other part, for example, to write the nextpagelink annotation. 

---
In reply to: [285801842](https://github.com/OData/odata.net/pull/1464#discussion_r285801842) [](ancestors = 285801842)

**Comment by xuzhg** [comment]
@mikepizzo https://github.com/OData/odata.net/issues/1472 is created for 7.6.1 update. If others are ok, can i merge this PR?

File: src/Microsoft.OData.Core/ODataWriterCore.cs [hunk 4]
```diff
@@ -1800,15 +1800,22 @@ private void WriteEntityReferenceLinkImplementation(ODataEntityReferenceLink ent
...
             {
                 this.InterceptException(() =>
                 {
                     WriterValidationUtils.ValidateEntityReferenceLink(entityReferenceLink);
-                    this.WriteEntityReferenceInNavigationLinkContent((ODataNestedResourceInfo)this.CurrentScope.Item, entityReferenceLink);
+
+                    ODataNestedResourceInfo nestedInfo = this.CurrentScope.Item as ODataNestedResourceInfo;
+                    if (nestedInfo == null)
+                    {
+                        nestedInfo = this.ParentNestedResourceInfoScope.Item as ODataNestedResourceInfo;
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/ODataWriterCore.cs]: >ParentNestedResourceInfoScope [](start = 42, length = 29)

Can we safely assert that ParentNestedResourceInfoScope will always be non-null?  What if the user tries to call WriteEntityReferenceLinkImplementation at the top level?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 3 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Yech. As discussed, we need to fix this to support the same pattern in request and response (the new one) and deprecate the old pattern.

---
In reply to: [289081395](https://github.com/OData/odata.net/pull/1464#discussion_r289081395) [](ancestors = 289081395,285803927)

**Comment by mikepizzo** [comment]
I'm okay merging once the remaining active comments are addressed, and fixing the case for writing requests in the new pattern in 7.6.1.

---
In reply to: [497859608](https://github.com/OData/odata.net/pull/1464#issuecomment-497859608) [](ancestors = 497859608)

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 2 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: Resolved.

---
In reply to: [289888907](https://github.com/OData/odata.net/pull/1464#discussion_r289888907) [](ancestors = 289888907,285800014)

File: src/Microsoft.OData.Core/ODataWriterCore.cs (hunk 4 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/ODataWriterCore.cs]: I think we can.

---
In reply to: [289887974](https://github.com/OData/odata.net/pull/1464#discussion_r289887974) [](ancestors = 289887974)

File: src/Microsoft.OData.Core/ODataWriterCore.cs [hunk 5]
```diff
@@ -1970,7 +1977,11 @@ private void CheckForNestedResourceInfoWithContent(ODataPayloadKind contentPaylo
             {
                 if (contentPayloadKind == ODataPayloadKind.EntityReferenceLink)
                 {
-                    this.ThrowODataException(Strings.ODataWriterCore_EntityReferenceLinkWithoutNavigationLink, null);
+                    Scope parenScope = this.ParentNestedResourceInfoScope;
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/ODataWriterCore.cs]: Yes. We can. 

---
In reply to: [289888445](https://github.com/OData/odata.net/pull/1464#discussion_r289888445) [](ancestors = 289888445)

**Comment by xuzhg** [comment]
Merged. Thanks.

**Pull request closed without merging**